### PR TITLE
Fix #62

### LINKED ISSUE #62: Dimension coordinate with auto-numbering gives error
If I provide `length` instead of `values` to `pymc.Model.add_chord`, I get the error below.
I assume this is a bug.

Kind regards,
Ivar Rummelhoff

```
TypeError                                 Traceback (most recent call last)
Cell In[6], line 2
      1 with model:
----> 2     idata = pm.sample(random_seed=9641478, nuts_sampler='nutpie')

File .venv/lib/python3.11/site-packages/pymc/sampling/mcmc.py:658, in sample(draws, tune, chains, cores, random_seed, progressbar, step, nuts_sampler, initvals, init, jitter_max_retries, n_init, trace, discard_tuned_samples, compute_convergence_checks, keep_warning_stat, return_inferencedata, idata_kwargs, nuts_sampler_kwargs, callback, mp_ctx, model, **kwargs)
    654     if not isinstance(step, NUTS):
    655         raise ValueError(
    656             "Model can not be sampled with NUTS alone. Your model is probably not continuous."
    657         )
--> 658     return _sample_external_nuts(
    659         sampler=nuts_sampler,
    660         draws=draws,
    661         tune=tune,
    662         chains=chains,
    663         target_accept=kwargs.pop("nuts", {}).get("target_accept", 0.8),
    664         random_seed=random_seed,
    665         initvals=initvals,
    666         model=model,
    667         progressbar=progressbar,
    668         idata_kwargs=idata_kwargs,
    669         nuts_sampler_kwargs=nuts_sampler_kwargs,
    670         **kwargs,
    671     )
    673 if isinstance(step, list):
    674     step = CompoundStep(step)

File .venv/lib/python3.11/site-packages/pymc/sampling/mcmc.py:297, in _sample_external_nuts(sampler, draws, tune, chains, target_accept, random_seed, initvals, model, progressbar, idata_kwargs, nuts_sampler_kwargs, **kwargs)
    291 if idata_kwargs is not None:
    292     warnings.warn(
    293         "`idata_kwargs` are currently ignored by the nutpie sampler",
    294         UserWarning,
    295     )
--> 297 compiled_model = nutpie.compile_pymc_model(model)
    298 idata = nutpie.sample(
    299     compiled_model,
    300     draws=draws,
   (...)
    306     **nuts_sampler_kwargs,
    307 )
    308 return idata

File .venv/lib/python3.11/site-packages/nutpie/compile_pymc.py:190, in compile_pymc_model(model, **kwargs)
    185 expand_numba_raw, c_sig_expand = _make_c_expand_func(
    186     n_dim, n_expanded, expand_fn, user_data, shared_expand, shared_data
    187 )
    188 expand_numba = numba.cfunc(c_sig_expand, **kwargs)(expand_numba_raw)
--> 190 coords = {name: pd.Index(vals) for name, vals in model.coords.items()}
    191 if "unconstrained_parameter" in coords:
    192     raise ValueError("Model contains invalid name 'unconstrained_parameter'.")

File .venv/lib/python3.11/site-packages/nutpie/compile_pymc.py:190, in <dictcomp>(.0)
    185 expand_numba_raw, c_sig_expand = _make_c_expand_func(
    186     n_dim, n_expanded, expand_fn, user_data, shared_expand, shared_data
    187 )
    188 expand_numba = numba.cfunc(c_sig_expand, **kwargs)(expand_numba_raw)
--> 190 coords = {name: pd.Index(vals) for name, vals in model.coords.items()}
    191 if "unconstrained_parameter" in coords:
    192     raise ValueError("Model contains invalid name 'unconstrained_parameter'.")

File .venv/lib/python3.11/site-packages/pandas/core/indexes/base.py:524, in Index.__new__(cls, data, dtype, copy, name, tupleize_cols)
    521         data = com.asarray_tuplesafe(data, dtype=_dtype_obj)
    523 elif is_scalar(data):
--> 524     raise cls._raise_scalar_data_error(data)
    525 elif hasattr(data, "__array__"):
    526     return Index(np.asarray(data), dtype=dtype, copy=copy, name=name)

File .venv/lib/python3.11/site-packages/pandas/core/indexes/base.py:5265, in Index._raise_scalar_data_error(cls, data)
   5260 @final
   5261 @classmethod
   5262 def _raise_scalar_data_error(cls, data):
   5263     # We return the TypeError so that we can raise it from the constructor
   5264     #  in order to keep mypy happy
-> 5265     raise TypeError(
   5266         f"{cls.__name__}(...) must be called with a collection of some "
   5267         f"kind, {repr(data) if not isinstance(data, np.generic) else str(data)} "
   5268         "was passed"
   5269     )

TypeError: Index(...) must be called with a collection of some kind, None was passed
```

### PR DESCRIPTION
**Pull request by aseyboldt**
Seems the coords can be None in pymc...
fixes #62

### CODE REVIEW

File: python/advi.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+@numba.njit(parallel=True)
+def compute_grads(draws):
+    out = np.empty_like(draws)
+    for i in numba.prange(len(draws)):
+        _, grad = logp_func(draws[i])
+        out[i] = grad
+    return out
+
+
+def advi(compiled_model, max_iter=10_000, residual=5e-3):
```
Comment (twiecki) [inline python/advi.py:32]: don't we have ADVI at the pytensor level in pymc already?

Comment (aseyboldt) [reply python/advi.py]: Sorry, this wasn't supposed to end up in the PR. I rebased

**Comment by twiecki** [comment]
```
_______________________________ test_pymc_model ________________________________

    def test_pymc_model():
        with pm.Model() as model:
            model.add_coord("foo", length=5)
            pm.Normal("a", dims="foo")
    
>       compiled = nutpie.compile_pymc_model(model)

tests/test_pymc.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/nutpie/compile_pymc.py:191: in compile_pymc_model
    coords = {name: pd.Index(vals) for name, vals in model.coords.items()}
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/nutpie/compile_pymc.py:191: in <dictcomp>
    coords = {name: pd.Index(vals) for name, vals in model.coords.items()}
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pandas/core/indexes/base.py:524: in __new__
    raise cls._raise_scalar_data_error(data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pandas.core.indexes.base.Index'>, data = None

    @final
    @classmethod
    def _raise_scalar_data_error(cls, data):
        # We return the TypeError so that we can raise it from the constructor
        #  in order to keep mypy happy
>       raise TypeError(
            f"{cls.__name__}(...) must be called with a collection of some "
            f"kind, {repr(data) if not isinstance(data, np.generic) else str(data)} "
            "was passed"
        )
E       TypeError: Index(...) must be called with a collection of some kind, None was passed

/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pandas/core/indexes/base.py:5265: TypeError
```

**Comment by twiecki** [comment]
```
Processing ./dist/nutpie-0.9.1-cp310-cp310-manylinux_2_28_x86_64.whl
INFO: pip is looking at multiple versions of nutpie[all] to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement pyarrow>=12.0.0 (from nutpie[all]) (from versions: none)
ERROR: No matching distribution found for pyarrow>=12.0.0
```

**Comment by maresb** [comment]
Not sure why this would fail on OSX but succeed on Linux:

```
=================================== FAILURES ===================================
_______________________________ test_pymc_model ________________________________

    def test_pymc_model():
        with pm.Model() as model:
            model.add_coord("foo", length=5)
            pm.Normal("a", dims="foo")
    
>       compiled = nutpie.compile_pymc_model(model)

tests/test_pymc.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/nutpie/compile_pymc.py:191: in compile_pymc_model
    coords = {name: pd.Index(vals) for name, vals in model.coords.items()}
../../../hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/nutpie/compile_pymc.py:191: in <dictcomp>
    coords = {name: pd.Index(vals) for name, vals in model.coords.items()}
../../../hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pandas/core/indexes/base.py:524: in __new__
    raise cls._raise_scalar_data_error(data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'pandas.core.indexes.base.Index'>, data = None

    @final
    @classmethod
    def _raise_scalar_data_error(cls, data):
        # We return the TypeError so that we can raise it from the constructor
        #  in order to keep mypy happy
>       raise TypeError(
            f"{cls.__name__}(...) must be called with a collection of some "
            f"kind, {repr(data) if not isinstance(data, np.generic) else str(data)} "
            "was passed"
        )
E       TypeError: Index(...) must be called with a collection of some kind, None was passed

../../../hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pandas/core/indexes/base.py:5265: TypeError
```

**Comment by aseyboldt** [comment]
Thanks for having a look!

The test failure is I think because it seems to be using the nutpie from pypi instead of the new one. I don't quite understand why that would happen on mac but not on linux though...

**Comment by aseyboldt** [comment]
Ah, I think we need the separate install step in the other ci jobs as well...

**Comment by aseyboldt** [comment]
> pip install --force-reinstall dist/nutpie*.whl

I think the problem with this is that there are several whl files, for different python versions.

**Pull request merged**